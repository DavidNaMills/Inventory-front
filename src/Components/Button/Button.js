import React from 'react';
import classes from './button.module.scss';

const allTypes = {
    base: {
        base: classes.button__base,
        shadow: classes.button__base_shadow,
        border: classes.button__base_border,
        skeleton: classes.button__base_skeleton,
    },

    submit: {
        base: classes.button__submit,
        shadow: classes.button__submit_shadow,
        border: classes.button__submit_border,
        skeleton: classes.button__submit_skeleton,
    },

    clear: {
        base: classes.button__clear,
        shadow: classes.button__clear_shadow,
        border: classes.button__clear_border,
        skeleton: classes.button__clear_skeleton,
    },

    danger: {
        base: classes.button__danger,
        shadow: classes.button__danger_shadow,
        border: classes.button__danger_border,
        skeleton: classes.button__danger_skeleton,
    },

    information: {
        base: classes.button__information,
        shadow: classes.button__information_shadow,
        border: classes.button__information_border,
        skeleton: classes.button__information_skeleton,
    },
};


const Button = ({ config = {}, onClick, label = null, skeleton = false, raised = false, full = false, type = 'base', border = false, bold = false, wide = false, isDisabled = false, Icon = null, noStyle=null }) => {
    const fullStyle = [classes.button];
    let which = type;
    const style = {};

    if (isDisabled) {
        fullStyle.push(classes.button__disabled);
    } else if(noStyle) { fullStyle.push(classes.button_noStyle);
    } else {
        if (allTypes[type]) {
            which = type;
            fullStyle.push(allTypes[type].base);
        } else {
            which = 'base';
            fullStyle.push(allTypes['base'].base);
        }

        if (skeleton) { fullStyle.push(allTypes[which].skeleton); }
        if (raised) { fullStyle.push(allTypes[which].shadow); }
        if (border) { fullStyle.push(allTypes[which].border); }
    }
    if (full) { fullStyle.push(classes.button_full); }
    if (bold) { fullStyle.push(classes.button_bold); }
    
    
    if (wide) {
        style.width = '30%';
    }

    return (
        <button
            {...config}
            onClick={onClick}
            style={style}
            className={fullStyle.join(' ')}
            disabled={isDisabled}
        ><p>{label}</p>{Icon && <Icon />}</button>
    )
}

export default React.memo(Button);