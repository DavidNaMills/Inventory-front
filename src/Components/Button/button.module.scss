@import '../../styles/mixins/buttonMixins.module.scss';
@import '../../styles/variables/colors.module.scss';

.button{
    display: flex;
    align-content: center;
    justify-content: center;
    font-size: 16px;
    border: none;
    padding: 10px;
    padding-left: 20px;
    padding-right: 20px;
    border-radius: 2px;

    p{
        padding-top: 5px;
        padding-right: 10px;
    }

    &__disabled{
        background-color: $disabled-back;
        color: $disabled-color;
    }

    &__base{
        @include buttonStyle($base-back, $base-color);
        &:hover{
            @include onHover($base-shadow, $light-text, $base-back);
        }

        &_shadow{@include buttonShadow($base-shadow);}
        &_border{@include buildBorder($base-color);}
        &_skeleton{@include buildSkeleton($base-back, $base-color); }
    };
    
    &__submit{
        @include buttonStyle($submit-back, $submit-color);
        &:hover{
            @include onHover($submit-shadow, $light-text, $submit-back);
        }

        &_shadow{ @include buttonShadow($submit-shadow); }
        &_border{ @include buildBorder($submit-color); }
        &_skeleton{@include buildSkeleton($submit-back, $submit-color); }
    }
    
    &__clear{
        @include buttonStyle($clear-back, $clear-color);
        &:hover{
            @include onHover($clear-shadow, $clear-color, $clear-back);
        }

        &_shadow{ @include buttonShadow($clear-shadow); }
        &_border{ @include buildBorder($clear-color); }
        &_skeleton{@include buildSkeleton($clear-back, $clear-color); }
    }
    
    &__danger{
        @include buttonStyle($danger-back, $danger-color);
        &:hover{
            @include onHover($danger-shadow, $light-text, $danger-back);
        }
        &_shadow{ @include buttonShadow($danger-shadow); }
        &_border{ @include buildBorder($danger-color); }
        &_skeleton{@include buildSkeleton($danger-back, $danger-color); }
    }
    
    &__information{
        @include buttonStyle($information-back, $information-color);
        &:hover{
            @include onHover($information-shadow, $light-text, $information-back);
        }
        &_shadow{ @include buttonShadow($information-shadow); }
        &_border{ @include buildBorder($information-color); }
        &_skeleton{@include buildSkeleton($information-back, $information-color); }
    }

    &_flat{
        border: none;
        box-shadow: none;
        -webkit-box-shadow: none;
        -moz-box-shadow: none;
    }

    &_full{
        width: 100%;
    }

    &_bold{
        font-weight: bold;
    }

    &_noStyle{
        background-color: transparent;
        border: none;
        box-shadow: none;

        &:hover{
            background-color: rgb(212, 212, 212);
            padding-top: 10px;
            padding-bottom: 10px;
            padding-left: 5px;
            padding-right: 10px;
            border-radius: 100px;
        }
    }
}