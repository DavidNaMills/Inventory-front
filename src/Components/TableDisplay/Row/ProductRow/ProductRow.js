import React from 'react';
import { isMobileOnly } from 'react-device-detect';
import classes from './ProductRow.module.scss';
import ImageDisplay from '../../../ImageDisplay/ImageDisplay';
import Title from '../../../Title/Title';
import { useLocale } from '../../../../Context/LocaleContext';

const ProductRow = (props) => {
    const { config, data, withTitle=true, withImage=true, withInfo=true, stack=false } = props;
    const locale = useLocale();

    return (
        <div className={classes.productRow}>
            {withImage && <ImageDisplay url={data.url ? `/productImages/${data.url}` : null} alt={`${data.name}-${data.code}`} size='medium'/>}
            <div className={classes.productRow__infoCont}>
                {withTitle && <Title message={data[config.main[0]]} size='h4' />}
                {withInfo &&
                    <div className={stack? classes.productRow__cellCont_stack : classes.productRow__cellCont}>
                        <p><span>{locale[config.main[1]]}:</span>{data[config.main[1]]}</p>
                        <p><span>{locale[config.main[2]]}:</span>{data[config.main[2]]}</p>
                    </div>
                }
            </div>
        </div>
    )
}

export default ProductRow;