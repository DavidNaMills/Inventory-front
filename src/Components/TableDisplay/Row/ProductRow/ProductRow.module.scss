.productRow{
    display: flex;
    align-items: center;
    
    &__infoCont{
        display: flex;
        flex-direction: column;
        margin-left: 20px;

    }
    
    

    &__cellCont{
        width: 100%;
        padding-top: 20px;
        display: flex;

        p{
            margin-right: 50px;
        }

        p:last-of-type {
            margin-right: none;
        }

    }

    &__cellCont_stack{
        width: 100%;
        padding-top: 10px;
        padding-bottom: 10px;
        display: flex;
        flex-direction: column;

        p{
            margin-right: 50px;
        }

        p:last-of-type {
            margin-right: none;
        }

    }



    span{
        font-weight: bold;
        margin-right: 10px;
    }
    
}
