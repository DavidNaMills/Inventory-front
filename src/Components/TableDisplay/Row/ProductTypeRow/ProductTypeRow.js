import React from 'react';
import classes from './ProductTypeRow.module.scss';
import Title from '../../../Title/Title';
import {isMobileOnly} from 'react-device-detect';
import { useLocale } from '../../../../Context/LocaleContext';

const ProductTypeRow = (props) => {
    const { config, data } = props;
    const locale = useLocale();

    return (
        <div className={classes.productTypeRow}>
            <div className={classes.productTypeRow__title}>
                <Title message={data[config.main[0]]} size='h4' />
            </div>
            {!isMobileOnly &&
                <div className={classes.productTypeRow__cellCont}>
                    <p><span>{locale[config.main[1]]}:</span>{data[config.main[1]]}</p>
                </div>
            }
        </div>
    )
}

export default ProductTypeRow;