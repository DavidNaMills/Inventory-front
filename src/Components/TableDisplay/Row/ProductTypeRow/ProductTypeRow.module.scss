.productTypeRow{
        // width: 75%;
        display: flex;
        align-items: baseline;
        // margin-left: 20px;
        padding: 10px;
        margin-bottom: 10px;

    &__title{
        margin-right: 50px;
    }

    &__cellCont{
        display:flex;
        padding-top: 20px;

        p{
            margin-right: 50px;
        }

        p:last-of-type {
            margin-right: none;
        }
    }
    
    &__cellContPortrait{
        display:flex;
        flex-direction: column;
        padding-top: 20px;

        p{
            margin-right: 50px;
        }

        p:last-of-type {
            margin-right: none;
        }
    }
    span{
        font-weight: bold;
        margin-right: 10px;
    }
}