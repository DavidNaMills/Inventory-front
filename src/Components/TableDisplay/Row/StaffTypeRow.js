import React from 'react';
import { isTablet, withOrientationChange } from 'react-device-detect';
import Title from '../../Title/Title';
import classes from './ProductTypeRow/ProductTypeRow.module.scss';
import { useLocale } from '../../../Context/LocaleContext';

const StaffTypeRow = (props) => {
    const locale = useLocale();
    const { config, data, withTitle = true, showInfo=true } = props;

    const isRole = (key, txt) =>{
        if(key === 'role'){
            return locale[txt];
        } else {
            return txt;
        }
    }

    return (
        <div className={classes.productTypeRow}>
            {withTitle &&
                <div className={classes.productTypeRow__title}>
                    <Title message={data[config.main[0]]} size='h3' />
                </div>
            }
            {showInfo &&
                <div className={(isTablet && props.isPortrait) ? classes.productTypeRow__cellContPortrait : classes.productTypeRow__cellCont}>
                    <p><span>{locale[config.main[1]]}:</span>{isRole(config.main[1], data[config.main[1]])}</p>
                    <p><span>{locale[config.main[2]]}:</span>{isRole(config.main[2], data[config.main[2]])}</p>
                </div>
            }
        </div>
    )
}

export default withOrientationChange(StaffTypeRow);