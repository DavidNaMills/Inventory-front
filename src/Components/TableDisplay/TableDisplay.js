import React from 'react';
import classes from './tableDisplay.module.scss';
import { useLocale } from '../../Context/LocaleContext';
import { isMobileOnly } from 'react-device-detect';
import StaffTypeRow from './Row/StaffTypeRow';
import ProductTypeRow from './Row/ProductTypeRow/ProductTypeRow';
import ProductRow from './Row/ProductRow/ProductRow';
import Button from '../Button/Button';
import MoreHorizIcon from '@material-ui/icons/MoreHoriz';
import CreateIcon from '@material-ui/icons/Create';


export const TableDisplay = (props) => {
    const locale = useLocale();
    const { list, config, update = null, remove = null, formName, viewMore = null, isSelected=null } = props;

    const all = list.map((x, count) => {
        let temp = [];
        const tStyle = [classes.tableDisplay];

        if (isSelected === x._id) { tStyle.push(classes.tableDisplay__selected) }
        else if (count % 2) { tStyle.push(classes.tableDisplay__even) }
        else { tStyle.push(classes.tableDisplay__odd) }

        switch (formName) {
            case 'staff':
                temp.push(
                    <StaffTypeRow
                        key={count}
                        config={config}
                        data={x}
                        showInfo={isMobileOnly ? false : true}
                    />
                )
                break;
            case 'productTypes':
                temp.push(
                    <ProductTypeRow
                        key={count}
                        config={config}
                        data={x}
                    />
                )
                break;

            case 'locationTypes':
                temp.push(
                    <ProductTypeRow
                        key={count}
                        config={config}
                        data={x}
                    />
                )
                break;
            case 'products':
                temp.push(
                    <ProductRow
                        key={count}
                        config={config}
                        data={x}
                        withImage={isMobileOnly? false : true}
                        withInfo={isMobileOnly? false : true}
                    />
                )
                break;
            default:
                return <p key={count}>ERROR</p>
        }
        return (
            <div className={tStyle.join(' ')} key={count}>
                {temp}
                <div className={classes.tableDisplay__optionsStyle}>
                    {update && temp.length > 0 &&
                        <Button onClick={() => update(JSON.parse(JSON.stringify(x)))} noStyle Icon={CreateIcon} />
                    }
                    {remove && temp.length > 0 && <Button onClick={() => remove(x._id)} type='danger' label={locale.remove} />}
                    {viewMore && <Button onClick={() => viewMore(x)} type='base' noStyle Icon={MoreHorizIcon} />}
                </div>
            </div>
        )
    })

    return (
        <div>
            { 
                all.length > 0
                ? all
                : <div className={classes.tableDisplay__message}><p className={classes.tableDisplay__message_text}>Nothing to display</p></div>
            }
            
        </div>
    );
}

export default TableDisplay;