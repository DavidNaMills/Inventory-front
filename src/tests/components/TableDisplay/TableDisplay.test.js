import React from 'react';
import '../../testHelpers/context';
import renderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';

import {TableDisplay} from '../../../Components/TableDisplay/TableDisplay';
import Button from '../../../Components/Button/Button';
import StaffTypeRow from '../../../Components/TableDisplay/Row/StaffTypeRow';
import ProductTypeRow from '../../../Components/TableDisplay/Row/ProductTypeRow/ProductTypeRow';
import ProductRow from '../../../Components/TableDisplay/Row/ProductRow/ProductRow';

import MoreHorizIcon from '@material-ui/icons/MoreHoriz';
import CreateIcon from '@material-ui/icons/Create';

const baseProps = {
    config: { main: ['name', 'code', 'qtyInStock'] },
    list: [
        {
            name: 'Blue dogs',
            location: 'Dundee',
            qtyInStock: '12345678',
            code: '187wertt56',
            url: 'test.jpg'
        }
    ]
}


describe('<TableDisplay /> component test suite', () => {
    it('matches snapshot', () => {
        const comp = renderer.create(<TableDisplay {...baseProps} formName='staff'/>);
        const app = comp.toJSON();
        expect(app).toMatchSnapshot();
    });

    it('exists', () => { 
        const wrapper = mount(<TableDisplay {...baseProps} formName='staff'/>);
        expect(wrapper.exists).toBeTruthy();
    });

    describe('Row component renders', () => {
        [
            { type: 'StaffTypeRow', form: 'staff', a: 0, b: 0, c: 1 },
            { type: 'ProductTypeRow', form: 'productTypes', a: 1, b: 0, c: 0 },
            { type: 'ProductTypeRow', form: 'locationTypes', a: 1, b: 0, c: 0 },
            { type: 'ProductRow', form: 'products', a: 0, b: 1, c: 0 },

            { type: 'StaffTypeRow', form: 'staff', a: 0, b: 0, c: 1 },
            { type: 'StaffTypeRow', form: 'staff', a: 0, b: 0, c: 1 },
        ]. forEach(x=>{

            it(`renders a <${x.type}/> component if formName = "${x.form}"`, () => { 
                const wrapper = mount(<TableDisplay {...baseProps} formName={x.form}/>);
                expect(wrapper.find(ProductTypeRow).length).toBe(x.a);
                expect(wrapper.find(ProductRow).length).toBe(x.b);
                expect(wrapper.find(StaffTypeRow).length).toBe(x.c);
                
                expect(wrapper.find(x.type).props()).toHaveProperty('config', baseProps.config);
                expect(wrapper.find(x.type).props()).toHaveProperty('data', baseProps.list[0]);
            });

        });
        
        it('renders a <p> element with text "ERROR"', () => { 
            const wrapper = mount(<TableDisplay {...baseProps} formName='invalid'/>);
            expect(wrapper.find('p').length).toBe(1);
            expect(wrapper.find('p').text()).toEqual('ERROR');

            expect(wrapper.find(StaffTypeRow).length).toBe(0);
            expect(wrapper.find(ProductTypeRow).length).toBe(0);
            expect(wrapper.find(ProductRow).length).toBe(0);
        });
    });

    describe('Styles and options', () => {
        it('renders 2 <div> elements with classNames: tableDisplay and tableDisplay__optionsStyle ', () => { 
            const wrapper = mount(<TableDisplay {...baseProps} formName='staff'/>);
            const div = wrapper.find('div');
            expect(div.length).toBe(7);
            expect(div.at(1).props()).toHaveProperty('className', 'tableDisplay tableDisplay__odd');
            expect(div.at(6).props()).toHaveProperty('className', 'tableDisplay__optionsStyle');
        });
        
        it('renders a Button if update is not null and there are at least 1 row', () => { 
            const t = jest.fn();
            const wrapper = mount(<TableDisplay {...baseProps} formName='staff' update={t} />);
            const btn = wrapper.find(Button);
            expect(btn.length).toBe(1);
            expect(btn.props()).toHaveProperty('noStyle', true);
            expect(btn.props()).toHaveProperty('Icon', CreateIcon);
        });
        
        it('does not render a Button if update is not null but there are 0 rows', () => { 
            const t = jest.fn();
            const wrapper = mount(<TableDisplay {...baseProps} list={[]} formName='staff' update={t} />);
            const btn = wrapper.find(Button);
            expect(btn.length).toBe(0);
        });

        it('renders a Button if remove is not null and there are at least 1 row', () => { 
            const t = jest.fn();
            const wrapper = mount(<TableDisplay {...baseProps} formName='staff' remove={t} />);
            const btn = wrapper.find(Button);
            expect(btn.length).toBe(1);
            expect(btn.props()).toHaveProperty('type', 'danger');
            expect(btn.props()).toHaveProperty('label', 'remove');
        });
        
        it('does not render a Button if remove is not null but there are 0 rows', () => { 
            const t = jest.fn();
            const wrapper = mount(<TableDisplay {...baseProps} list={[]} formName='staff' remove={t} />);
            const btn = wrapper.find(Button);
            expect(btn.length).toBe(0);
        });


        it('renders a Button if viewMore is not null and there are at least 1 row', () => {
            const t = jest.fn();
            const wrapper = mount(<TableDisplay {...baseProps} formName='staff' viewMore={t} />);
            const btn = wrapper.find(Button);
            expect(btn.length).toBe(1);
            expect(btn.props()).toHaveProperty('type', 'base');
            expect(btn.props()).toHaveProperty('noStyle', true);
            expect(btn.props()).toHaveProperty('Icon', MoreHorizIcon);
         });
        
        it('does not render a Button if viewMore is not null but there are 0 rows', () => {
            const t = jest.fn();
            const wrapper = mount(<TableDisplay {...baseProps} formName='staff' list={[]} viewMore={t} />);
            const btn = wrapper.find(Button);
            expect(btn.length).toBe(0);
         });
    });
});