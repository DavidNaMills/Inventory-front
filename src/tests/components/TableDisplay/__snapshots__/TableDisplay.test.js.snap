// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableDisplay /> component test suite matches snapshot 1`] = `
<div>
  <div
    className="tableDisplay tableDisplay__odd"
  >
    <div
      className="productTypeRow"
    >
      <div
        className="productTypeRow__title"
      >
        <div>
          <h3
            className="title_h3 "
          >
            Blue dogs
          </h3>
        </div>
      </div>
      <div
        className="productTypeRow__cellCont"
      >
        <p>
          <span>
            Code
            :
          </span>
          187wertt56
        </p>
        <p>
          <span>
            Qty
            :
          </span>
          12345678
        </p>
      </div>
    </div>
    <div
      className="tableDisplay__optionsStyle"
    />
  </div>
</div>
`;
